File: src/push.ts
// This file is based on https://github.com/s0/git-publish-subdir-action
// MIT License
//
// Copyright (c) 2018 Sam Lanning
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import { stream as fgStream } from 'fast-glob';
import { promises as fs } from 'fs';
import { tmpdir } from 'os';
import * as path from 'path';
import cpy from 'cpy';

import { mkdirP } from '@actions/io';
import {
  add,
  checkout,
  clone,
  commit,
  configureEmail,
  configureName,
  fetchOrigin,
  hasBranch,
  push,
  rebase,
} from './git';
import { log } from './logger';
import { CompareOutput } from './compare';
import { workspace } from './path';
import * as constants from './constants';
import { backOff } from 'exponential-backoff';

export type PushImagesInput = {
  githubToken: string;
  runId: number;
  result: CompareOutput;
  branch: string;
  targetDir: string;
  env: EnvironmentVariables;
  retentionDays: number;
  // commitName?: string;
  // commitEmail?: string;
};

export interface EnvironmentVariables {
  GITHUB_REPOSITORY?: string;
  GITHUB_EVENT_PATH?: string;
  /** The name of the person / app that that initiated the workflow */
  GITHUB_ACTOR?: string;
}

declare global {
  namespace NodeJS {
    interface ProcessEnv extends EnvironmentVariables {}
  }
}

export type Config = {
  branch: string;
  repo: string;
};

/**
 * The GitHub event that triggered this action
 */
export type Event = {
  pusher?: {
    email?: string;
    name?: string;
  };
};

const genConfig = (input: PushImagesInput): Config => {
  const { branch, env } = input;
  // Determine the type of URL
  if (!input.githubToken) throw new Error('GITHUB_TOKEN must be specified when REPO == self');
  if (!env.GITHUB_REPOSITORY) throw new Error('GITHUB_REPOSITORY must be specified when REPO == self');

  const url = `https://x-access-token:${input.githubToken}@github.com/${env.GITHUB_REPOSITORY}.git`;
  const config: Config = {
    repo: url,
    branch,
  };
  return config;
};

const copyImages = async (result: CompareOutput, temp: string, dest: string): Promise<void> => {
  log.info(`Copying all files`);

  if (result.deletedItems.length > 0) {
    log.info(`Copying deleted files`);
    const deleted = result.deletedItems.map(item => `${path.join(workspace(), constants.EXPECTED_DIR_NAME)}/${item}`);
    await cpy(deleted, `${temp}/${dest}/expected/`);
  }

  if (result.newItems.length > 0) {
    log.info(`Copying new files`);
    const newGlobs = result.newItems.map(item => `${path.join(workspace(), constants.ACTUAL_DIR_NAME)}/${item}`);
    await cpy(newGlobs, `${temp}/${dest}/actual/`);
  }

  if (result.failedItems.length > 0) {
    const failedGlobs = result.failedItems.map(item => `${path.join(workspace(), constants.DIFF_DIR_NAME)}/${item}`);
    await cpy(failedGlobs, `${temp}/${dest}/diff/`);

    const expectedGlobs = result.failedItems.map(
      item => `${path.join(workspace(), constants.EXPECTED_DIR_NAME)}/${item}`,
    );
    await cpy(expectedGlobs, `${temp}/${dest}/expected/`);
    const actualGlobs = result.failedItems.map(item => `${path.join(workspace(), constants.ACTUAL_DIR_NAME)}/${item}`);
    await cpy(actualGlobs, `${temp}/${dest}/actual/`);
  }
  return;
};

export const pushImages = async (input: PushImagesInput) => {
  log.info('Staring Push images');
  const { env } = input;
  const config = genConfig(input);

  const TMP_PATH = await fs.mkdtemp(path.join(tmpdir(), 'reg-actions-'));
  const REPO_TEMP = path.join(TMP_PATH, 'repo');

  if (!env.GITHUB_EVENT_PATH) throw new Error('Expected GITHUB_EVENT_PATH');

  const event: Event = JSON.parse((await fs.readFile(env.GITHUB_EVENT_PATH)).toString());

  const name = /* input.commitName ?? */ event.pusher?.name ?? env.GITHUB_ACTOR ?? 'Git Publish Subdirectory';
  const email =
    // input.commitEmail ??
    event.pusher?.email ?? (env.GITHUB_ACTOR ? `${env.GITHUB_ACTOR}@users.noreply.github.com` : 'nobody@nowhere');

  // Set Git Config
  await configureName(name);
  await configureEmail(email);

  // Environment to pass to children
  const execEnv = env as { [key: string]: string };

  // Clone the target repo
  await clone({ repo: config.repo, dist: REPO_TEMP }, { env: execEnv });

  const execOptions = { env: execEnv, cwd: REPO_TEMP };

  // Fetch branch if it exists
  await fetchOrigin({ branch: config.branch }, execOptions).catch(err => {
    const s = err.toString();
    if (s.indexOf("Couldn't find remote ref") === -1) {
      log.warn("Failed to fetch target branch, probably doesn't exist");
      log.error(err);
    }
  });

  // Check if branch already exists
  log.info(`Checking if branch ${config.branch} exists already`);

  const branchExist = await hasBranch(config.branch, execOptions);

  if (!branchExist) {
    // Branch does not exist yet, let's check it out as an orphan
    log.info(`${config.branch} does not exist, creating as orphan`);
    await checkout(config.branch, true, execOptions);
  } else {
    await checkout(config.branch, false, execOptions);
  }

  // Update contents of branch
  log.info(`Updating branch ${config.branch}`);

  /**
   * The list of globs we'll use for clearing
   */

  if (!branchExist) {
    const globs = ['**/*', '!.git'];
    log.info(`Removing all files from target branch`);
    const filesToDelete = fgStream(globs, { absolute: true, dot: true, followSymbolicLinks: false, cwd: REPO_TEMP });
    log.info(filesToDelete);
    // Delete all files from the filestream
    for await (const entry of filesToDelete) {
      await fs.unlink(entry);
    }
  }

  /* delete expired directories */
  try {
    log.info(`Retention days = ${input.retentionDays}`);
    const retention = input.retentionDays * 24 * 60 * 60 * 1000;
    const files = await fs.readdir(REPO_TEMP);
    for (const fileOrDir of files) {
      const p = path.join(REPO_TEMP, fileOrDir);
      if ((await fs.stat(p)).isDirectory()) {
        const day = fileOrDir.split('_')[0];
        if (day) {
          const now = new Date().getTime();
          const target = new Date(day).getTime();
          if (now - target > retention) {
            log.info(`delete dir ${fileOrDir}`);
            await fs.rmdir(p, { recursive: true });
          }
        }
      }
    }
  } catch (e) {
    log.error('Failed to delete directories', e);
  }

  const destDir = input.targetDir;

  // Make sure the destination sourceDir exists
  await mkdirP(path.resolve(REPO_TEMP, destDir));

  await copyImages(input.result, REPO_TEMP, destDir);

  await add(execOptions);

  const message = `Update ${input.branch} to output generated at runId:${input.runId}`;
  await commit(message, execOptions);

  log.info(`Pushing`);

  (async () => {
    return backOff(
      async () => {
        if (branchExist) {
          await rebase(config.branch, execOptions);
        }
        const res = await push(config.branch, execOptions);
        log.info(res.stdout);
        log.info(`Deployment Successful`);
      },
      { numOfAttempts: 5 },
    );
  })();
};

File: src/push.rs
//! 比較結果画像を artifact ブランチに push する (`src/push.ts` の Rust 移植)。

use std::fs;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use anyhow::{anyhow, Context, Result};
use backon::{ExponentialBuilder, Retryable};
use serde::Deserialize;

use crate::compare::CompareOutput;
use crate::constants::{ACTUAL_DIR_NAME, DIFF_DIR_NAME, EXPECTED_DIR_NAME};
use crate::git;
use crate::path::workspace;

#[derive(Debug)]
pub struct PushImagesInput<'a> {
    pub github_token: &'a str,
    pub run_id: u64,
    pub result: &'a CompareOutput,
    pub branch: &'a str,
    pub target_dir: &'a str,
    pub retention_days: u32,
}

#[derive(Debug, Default, Deserialize)]
struct PusherEvent {
    pusher: Option<Pusher>,
}

#[derive(Debug, Default, Deserialize)]
struct Pusher {
    name: Option<String>,
    email: Option<String>,
}

pub async fn push_images(input: PushImagesInput<'_>) -> Result<()> {
    tracing::info!("starting push_images");
    let repo_env =
        std::env::var("GITHUB_REPOSITORY").map_err(|_| anyhow!("GITHUB_REPOSITORY must be set"))?;
    let actor = std::env::var("GITHUB_ACTOR").ok();
    let event_path =
        std::env::var("GITHUB_EVENT_PATH").map_err(|_| anyhow!("GITHUB_EVENT_PATH must be set"))?;
    let event: PusherEvent = serde_json::from_str(
        &fs::read_to_string(&event_path).with_context(|| format!("read {event_path}"))?,
    )
    .unwrap_or_default();

    let pusher = event.pusher.unwrap_or_default();
    let name = pusher
        .name
        .or_else(|| actor.clone())
        .unwrap_or_else(|| "Git Publish Subdirectory".to_string());
    let email = pusher.email.unwrap_or_else(|| {
        actor
            .as_ref()
            .map(|a| format!("{a}@users.noreply.github.com"))
            .unwrap_or_else(|| "nobody@nowhere".to_string())
    });

    git::configure_name(&name)?;
    git::configure_email(&email)?;

    let repo_url = format!(
        "https://x-access-token:{}@github.com/{repo_env}.git",
        input.github_token
    );

    let tmp = make_tmp_dir()?;
    let repo_temp = tmp.join("repo");
    git::clone(&repo_url, &repo_temp)?;

    // fetch (失敗してもブランチが無いだけかもしれないので警告のみ)
    if let Err(e) = git::fetch_origin(input.branch, &repo_temp) {
        let s = e.to_string();
        if !s.contains("Couldn't find remote ref") {
            tracing::warn!(error = %e, "fetch_origin failed");
        }
    }

    let branch_exists = git::has_branch(input.branch, &repo_temp)?;
    if branch_exists {
        git::checkout(input.branch, false, &repo_temp)?;
    } else {
        tracing::info!(branch = input.branch, "creating orphan branch");
        git::checkout(input.branch, true, &repo_temp)?;
        // orphan ブランチでは既存ファイルを全削除 (.git は除く)
        wipe_dir(&repo_temp, &[".git"])?;
    }

    // retention: 古いディレクトリ削除
    let retention = Duration::from_secs(input.retention_days as u64 * 24 * 60 * 60);
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO);
    if let Ok(entries) = fs::read_dir(&repo_temp) {
        for entry in entries.flatten() {
            let p = entry.path();
            if !p.is_dir() {
                continue;
            }
            let Some(name) = p.file_name().and_then(|s| s.to_str()) else {
                continue;
            };
            let Some(date_str) = name.split('_').next() else {
                continue;
            };
            // YYYY-MM-DD の3区切りだけを対象
            let parts: Vec<&str> = date_str.split('-').collect();
            if parts.len() != 3 {
                continue;
            }
            let Ok(parsed) = chrono::NaiveDate::parse_from_str(date_str, "%Y-%m-%d") else {
                continue;
            };
            let target = parsed
                .and_hms_opt(0, 0, 0)
                .map(|d| d.and_utc().timestamp())
                .unwrap_or(0) as u64;
            if now.as_secs().saturating_sub(target) > retention.as_secs() {
                tracing::info!(dir = name, "deleting expired dir");
                if let Err(e) = fs::remove_dir_all(&p) {
                    tracing::warn!(error = %e, dir = ?p, "remove_dir_all failed");
                }
            }
        }
    }

    let dest_dir = repo_temp.join(input.target_dir);
    fs::create_dir_all(&dest_dir).context("mkdir dest_dir")?;

    copy_images(input.result, &dest_dir)?;

    git::add_all(&repo_temp)?;
    git::commit(
        &format!(
            "Update {} to output generated at runId:{}",
            input.branch, input.run_id
        ),
        &repo_temp,
    )?;

    let push_branch = input.branch.to_string();
    let backoff = ExponentialBuilder::default()
        .with_max_times(5)
        .with_min_delay(Duration::from_millis(500));
    let cwd = repo_temp.clone();
    (|| async {
        if branch_exists {
            git::rebase(&push_branch, &cwd).context("rebase")?;
        }
        let r = git::push(&push_branch, &cwd).context("push")?;
        if r.code != 0 {
            anyhow::bail!("git push failed: {}", r.stderr.trim());
        }
        tracing::info!(stdout = %r.stdout.trim(), "deployment successful");
        Ok::<(), anyhow::Error>(())
    })
    .retry(backoff)
    .await?;

    Ok(())
}

fn make_tmp_dir() -> Result<PathBuf> {
    let p = std::env::temp_dir().join(format!(
        "reg-actions-{}",
        std::time::SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0)
    ));
    fs::create_dir_all(&p).context("mkdir tmp")?;
    Ok(p)
}

fn wipe_dir(dir: &Path, keep: &[&str]) -> Result<()> {
    for entry in fs::read_dir(dir)?.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if keep.iter().any(|k| *k == name_str) {
            continue;
        }
        let p = entry.path();
        if p.is_dir() {
            fs::remove_dir_all(&p).with_context(|| format!("rm -rf {}", p.display()))?;
        } else {
            fs::remove_file(&p).with_context(|| format!("rm {}", p.display()))?;
        }
    }
    Ok(())
}

fn copy_images(result: &CompareOutput, dest_root: &Path) -> Result<()> {
    let ws = workspace();

    if !result.deleted_items.is_empty() {
        let dest = dest_root.join("expected");
        fs::create_dir_all(&dest).ok();
        for item in &result.deleted_items {
            let src = ws.join(EXPECTED_DIR_NAME).join(item);
            copy_with_parents(&src, &dest.join(file_name(item)))?;
        }
    }
    if !result.new_items.is_empty() {
        let dest = dest_root.join("actual");
        fs::create_dir_all(&dest).ok();
        for item in &result.new_items {
            let src = ws.join(ACTUAL_DIR_NAME).join(item);
            copy_with_parents(&src, &dest.join(file_name(item)))?;
        }
    }
    if !result.failed_items.is_empty() {
        let diff_dest = dest_root.join("diff");
        let exp_dest = dest_root.join("expected");
        let act_dest = dest_root.join("actual");
        fs::create_dir_all(&diff_dest).ok();
        fs::create_dir_all(&exp_dest).ok();
        fs::create_dir_all(&act_dest).ok();
        for item in &result.failed_items {
            let webp = item_to_webp(item);
            copy_with_parents(
                &ws.join(DIFF_DIR_NAME).join(&webp),
                &diff_dest.join(file_name(&webp)),
            )?;
            copy_with_parents(
                &ws.join(EXPECTED_DIR_NAME).join(item),
                &exp_dest.join(file_name(item)),
            )?;
            copy_with_parents(
                &ws.join(ACTUAL_DIR_NAME).join(item),
                &act_dest.join(file_name(item)),
            )?;
        }
    }
    Ok(())
}

fn file_name(p: &str) -> String {
    Path::new(p)
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_else(|| p.to_string())
}

fn item_to_webp(p: &str) -> String {
    let path = Path::new(p);
    let stem = path
        .file_stem()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_default();
    let parent = path
        .parent()
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_default();
    if parent.is_empty() {
        format!("{stem}.webp")
    } else {
        format!("{parent}/{stem}.webp")
    }
}

fn copy_with_parents(src: &Path, dst: &Path) -> Result<()> {
    if let Some(parent) = dst.parent() {
        fs::create_dir_all(parent).ok();
    }
    if !src.exists() {
        tracing::warn!(src = ?src, "skip missing source file");
        return Ok(());
    }
    fs::copy(src, dst).with_context(|| format!("copy {} -> {}", src.display(), dst.display()))?;
    Ok(())
}
